### PR TITLE
fix: allow next.js apps to import global styles

### LINKED ISSUE #22525: Cypress 10 | SASS Global Style File Imports Within Next.js - Unable To Run Cypress Component Test
### Current behavior

After launching Cypress Component Test application for the `Next.js` project, and importing multiple SASS (`.scss`) files from within the `cypress/component.ts` file, then i am unable to run a defined component test within the file `news-aggregator-article.cy.tsx` (link to [file](https://github.com/Sayvai/nextjs-playground/blob/cypress-integration/components/news-aggregator/news-aggregator-article.cy.tsx) on repo). On the test rendering panel, only the text `Cannot GET /__cypress/src/index.html` is rendered out.

See the highlighted screenshot below:
cypress-10_import-scss_unable-to-run-component-test

Is it not currently possible to import / support SASS style files for component testing? Or am i missing a step? 🤔 

If SASS imports are currently not supported, are you able to advise or recommend how I may be able to render the styles within the component test?

> **Warning** - The component test will successfully run (without SASS styles) if the imported SASS (`.scss`) files are removed from within the `cypress/component.ts` file.

> **Note** - This is a Next.js project that was initially scaffolded via `npx create-next-app`, and then configured to use SASS afterwards (`npm install --save-dev sass`).

### Desired behavior

The component test should run within the Cypress Component Test application window, as well as successfully importing and rendering the styles defined within the SASS files (`.scss`)

### Test code to reproduce

You may reproduce the issue by cloning the project from the repo; [nextjs-playground](https://github.com/Sayvai/nextjs-playground), and then checkout the branch; `cypress-integration`

The relevant code files are listed below.

file: `component.ts`:

```js
...
import '../../styles/globals.scss'
import '../../styles/Index.module.scss'
import '../../components/news-aggregator/news-aggregator-article.module.scss'
...
```
file: `news-aggregator-article.cy.tsx` (component test):

```js
import NewsAggregatorArticle from "./news-aggregator-article"

describe('news-aggregator-article.tsx', () => {
  it('should render with article data', () => {
    // given
    const article = {
      id: 15547,
      title: 'InSight teams push for more science in lander’s final months, Curiosity continues trek around Gale Crater',
      url: 'https://www.nasaspaceflight.com/2022/06/insight-curiosity-june-2022/',
      imageUrl: 'https://www.nasaspaceflight.com/wp-content/uploads/2022/06/jpegPIA25287-1170x702.jpeg',
      summary: 'In May 2022, members of NASA’s InSight Mars lander team announced that power levels on the lander were diminishing and that they expected the lander to become inoperative by December 2022. However, InSight’s teams want to push the lander to do as much as possible in its final months and have chosen to operate the lander’s seismometer for longer than previously planned.',
      publishedAt: '2022-06-25T16:42:33.000Z',
      updatedAt: '2022-06-25T19:27:56.825Z',
      newsSite: 'Space Fight News',
      featured: false,
      launches: [
        {
            id: 'df8d4fdb-9add-4ce9-9f0e-aae1c61df212',
            provider: 'Launch Library 2'
        },
        {
            id: '3b633f7e-fad9-480e-9011-f38d1bb2f932',
            provider: 'Launch Library 2'
        }
      ],
      events: [],
  }

    cy.mount(<NewsAggregatorArticle article={article} />)
  })
})
```

### Cypress Version

10.2.0

### Other

_No response_

### PR DESCRIPTION
**Pull request by ZachJW34**
- Closes #22525

### User facing changelog
Allow Next.js apps to import global styles

### Additional details
Next.js restricts the imports of global css/scss files to the top-level `<App />` component. The Webpack configuration for these rules can be found [here](https://github.com/vercel/next.js/blob/canary/packages/next/build/webpack/config/blocks/css/index.ts). This PR removes the limitations that Next applies to global styles, allowing their import in the support file.

### Steps to test
I tweaked the `next` project fixture to include global style imports. If you remove the call to [allowGlobalStylesImports](https://github.com/cypress-io/cypress/blob/7f1dc8b8c057d69ec23ad4c89264161beb3f4c29/npm/webpack-dev-server/src/helpers/nextHandler.ts#L18) and rebuild the `@cypress/webpack-dev-server package`, the `npm/webpack-dev-server/cypress/e2e/next.cy.ts` system tests will fail.

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: system-tests/project-fixtures/next/cypress/support/component.ts [hunk 1]
```diff
@@ -1 +1,4 @@
-//
+// Importing global styles fails with Next.js due to restrictions on style imports.
```
Comment (lmiller1990) [inline system-tests/project-fixtures/next/cypress/support/component.ts:1]: Do we want to add some kind of assertion that the global style is actually applied to some element?

**Review by lmiller1990 — COMMENTED** [review]
One comment!

**Comment by Sayvai** [comment]
@ZachJW34 - I tried to install the binary (to test your changes) built by Circle CI from this PR via the [provided](https://github.com/cypress-io/cypress/pull/22769#commitcomment-78426895) command:

```bash
npm install https://cdn.cypress.io/beta/npm/10.3.1/linux-x64/zachw/issue-22525-a049ed7f3a2f9a709e466f60bf2da775312af521/cypress.tgz
```

But the install failed to install on my local [nextjs-playground](https://github.com/Sayvai/nextjs-playground) project with a `404` (not found) response. Click to expand the following error output below:

<details>
<summary>NPM install Cypress error output</summary>

```bash
$ npm install https://cdn.cypress.io/beta/npm/10.3.1/linux-x64/zachw/issue-22525-a049ed7f3a2f9a709e466f60bf2da775312af521/cypress.tgz

npm ERR! code 1
npm ERR! path /Users/sayvai/repos/Sayvai/nextjs-playground/node_modules/cypress
npm ERR! command failed
npm ERR! command sh -c node index.js --exec install
npm ERR! ⚠ Warning: You are installing a pre-release build of Cypress.
npm ERR! 
npm ERR! Bugs may be present which do not exist in production builds.
npm ERR! 
npm ERR! This build was created from:
npm ERR!   * Commit SHA: a049ed7f3a2f9a709e466f60bf2da775312af521
npm ERR!   * Commit Branch: zachw/issue-22525
npm ERR!   * Commit Timestamp: 2022-07-13T19:10:51.000Z
npm ERR! 
npm ERR! Installing Cypress (version: https://cdn.cypress.io/beta/binary/10.3.1/darwin-x64/zachw/issue-22525-a049ed7f3a2f9a709e466f60bf2da775312af521/cypress.zip)
npm ERR! 
npm ERR! [STARTED] Task without title.
npm ERR! The Cypress App could not be downloaded.
npm ERR! 
npm ERR! Does your workplace require a proxy to be used to access the Internet? If so, you must configure the HTTP_PROXY environment variable before downloading Cypress. Read more: https://on.cypress.io/proxy-configuration
npm ERR! 
npm ERR! Otherwise, please check network connectivity and try again:
npm ERR! 
npm ERR! ----------
npm ERR! 
npm ERR! URL: https://cdn.cypress.io/beta/binary/10.3.1/darwin-x64/zachw/issue-22525-a049ed7f3a2f9a709e466f60bf2da775312af521/cypress.zip
npm ERR! Error: Failed downloading the Cypress binary.
npm ERR! Response code: 404
npm ERR! Response message: Not Found
npm ERR! 
npm ERR! ----------
npm ERR! 
npm ERR! Platform: darwin-x64 (21.5.0)
npm ERR! Cypress Version: 10.3.1
npm ERR! [FAILED] The Cypress App could not be downloaded.
npm ERR! [FAILED] 
npm ERR! [FAILED] Does your workplace require a proxy to be used to access the Internet? If so, you must configure the HTTP_PROXY environment variable before downloading Cypress. Read more: https://on.cypress.io/proxy-configuration
npm ERR! [FAILED] 
npm ERR! [FAILED] Otherwise, please check network connectivity and try again:
npm ERR! [FAILED] 
npm ERR! [FAILED] ----------
npm ERR! [FAILED] 
npm ERR! [FAILED] URL: https://cdn.cypress.io/beta/binary/10.3.1/darwin-x64/zachw/issue-22525-a049ed7f3a2f9a709e466f60bf2da775312af521/cypress.zip
npm ERR! [FAILED] Error: Failed downloading the Cypress binary.
npm ERR! [FAILED] Response code: 404
npm ERR! [FAILED] Response message: Not Found
npm ERR! [FAILED] 
npm ERR! [FAILED] ----------
npm ERR! [FAILED] 
npm ERR! [FAILED] Platform: darwin-x64 (21.5.0)
npm ERR! [FAILED] Cypress Version: 10.3.1

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/sayvai/.npm/_logs/2022-07-14T07_58_33_424Z-debug-0.log
```
</details>

Also, when i looked further into the CI Circle workflow build pipeline for the [Build the Cypress binary](https://app.circleci.com/pipelines/github/cypress-io/cypress/40448/workflows/e5e3fc14-4061-4b80-ab0e-02ada6db9852/jobs/1664916) job, although it may have appeared successful, the underlying console output appears to error out on the final build step.

I noticed some workflow checks on this PR failed, and so not sure whether those failed checks relates to the build failure? 🤔

File: system-tests/project-fixtures/next/cypress/support/component.ts (hunk 1 above)
Comment (ZachJW34) [reply system-tests/project-fixtures/next/cypress/support/component.ts:1]: Added an assertion here: b0e8bf010a648f4ceb44a9c2d77c4d5aec09caba

**Comment by ZachJW34** [comment]
@Sayvai Not sure why the binary isn't loading. I've pushed another commit and will see if I can figure out what's wrong when the new binary is published. The logs at the end of the binary build job are misleading. CI runs the binary against a test that is expected to fail so it prints some expected error messages

**Review by lmiller1990 — APPROVED** [review]
I did not test on my machine yet but my comments are addressed, so going to approve. I can give it a try as well.

**Comment by Sayvai** [comment]
> @Sayvai Not sure why the binary isn't loading. I've pushed another commit and will see if I can figure out what's wrong when the new binary is published. The logs at the end of the binary build job are misleading. CI runs the binary against a test that is expected to fail so it prints some expected error messages

@ZachJW34 - Yeah, i am still having trouble installing the build generated from this PR with the same console error output (described in earlier [comment](https://github.com/cypress-io/cypress/pull/22769#issuecomment-1184136478)).

My theory is that it could be a private registry permissions auth issue, or i need some special localised npm configuration to access builds from third-party (non-default) registry, albeit via the Cypress CDN 🤔

**Comment by Sayvai** [comment]
@ZachJW34 - I've finally managed to download the build via the cypress-bot provided npm install command / URL. Thanks for fixing the build!

Now, when i re-ran the test with the PR build, the Cypress browser window still renders out the message `Cannot GET /__cypress/src/index.html`, preventing a run of the test. No change.  (see attached screenshot below)

Cypress-10 3 1-PR-Dev-Build-Error

And if i comment out the global CSS import from within `component.ts`, I am able to run the test, just without the global styles. Best to kill the Cypress process for reliable reproducibility.

Am i missing another step (configuration) with this change? 🤔

**Comment by ZachJW34** [comment]
I checked out your repo and the `cypress-integration` branch. After installing the new binary, I was able to run the tests and verify the global styles were applying. Commenting in and out the `import "../../styles/globals.scss";` live-reloaded and ran as expected.

Did you happen to install the new binary when you had an active Cypress instance? If not, can you give post a little more info on how I can reproduce what you are seeing?

**Comment by Sayvai** [comment]
@ZachJW34 - I installed the new binary after uninstalling the stable binary.

Here's the reproduction steps i undertook:

- Rename already established repo to in same directory (command: `mv repos/Sayvai/nextjs-playground repos/Sayvai/nextjs-playground-1`).

- Clone the repo again to start afresh (command: `git clone git@github.com:Sayvai/nextjs-playground.git`)

- `cd` into new repo

- checkout the branch `cypress-integration` (command: `git checkout cypress-integration`)

- uninstall stable Cypress binary (command: `npm uninstall -D cypress`)

- install the PR-generated binary (command: `npm install -D https://cdn.cypress.io/beta/npm/10.3.1/darwin-x64/zachw/issue-22525-0cb8f895934845abb6d6006b21a59ee3aefdb3d5/cypress.tgz`)

- install all other dependencies (command: `npm install`)

- copy the certificate SSL keys from renamed project to newly cloned project (commands: `cp ../nextjs-playground-1/localhost.pem ./localhost.pem` and `cp ../nextjs-playground-1/localhost-key.pem ./localhost-key.pem`)

- run main project application (command: `npm run dev`)

- run Cypress interactively in a new terminal tab (command: `npm run cypress:open`)

- the Cypress electron application window opens

- click on the `Component Testing` box within the Cypress electron application window

- notice within the terminal output for Cypress, 2 `No template for dependency: ModuleHotAcceptDependency` errors are logged out (expand console output below; "Cypress Terminal Output"). First error refers to the non-gloabl SCSS import inside an actual component test file of [news-aggregator-article.cy.tsx](https://github.com/Sayvai/nextjs-playground/blob/cypress-integration/components/news-aggregator/news-aggregator-article.cy.tsx#L2), and the second error refers to the global SCSS import within the [component.ts](https://github.com/Sayvai/nextjs-playground/blob/cypress-integration/cypress/support/component.ts#L19) config file.

<details>
<summary><em>Cypress Terminal Output<em></summary>

```bash
$ npm run cypress:open

> nextjs-playground@0.1.0 cypress:open
> ./node_modules/.bin/cypress open

Missing baseUrl in compilerOptions. tsconfig-paths will be skipped
[@cypress/webpack-dev-server]: removing HotModuleReplacementPlugin from configuration.
<i> [webpack-dev-server] Project is running at:
<i> [webpack-dev-server] Loopback: http://127.0.0.1:8080/
<i> [webpack-dev-server] Content not from webpack is served from '/Users/sayvai/repos/Sayvai/nextjs-playground/public' directory
13 assets
137 modules

ERROR in ./components/layout/layout.module.scss
No template for dependency: ModuleHotAcceptDependency

ERROR in ./styles/globals.scss
No template for dependency: ModuleHotAcceptDependency

client (webpack 5.73.0) compiled with 2 errors in 868 ms
```
</details>

- now proceed to start to run the component test `news-aggregator-article.cy.tsx`

- the Cypress test browser window fails to run the test, and instead outputs onto the page `Cannot GET /__cypress/src/index.html`

Have a gone wrong in any of the above steps? Are you able to reproduce the error i observed? 🤔

**Comment by lmiller1990** [comment]
I had the **exact** same result as @ZachJW34 - down to the "comment in, comment out" part, [see my post here](https://github.com/cypress-io/cypress/issues/22525#issuecomment-1179851048). This weird behavior was **without** this branch. I am not sure why that is happening. I did not try on this branch, yet.

**Comment by ZachJW34** [comment]
@Sayvai thanks for the thorough reproduction. Unfortunately, I can't reproduce this issue following your steps. @lmiller1990 when you pull this down, can you go through the reproduction steps provided by @Sayvai and see if you are experiencing any issues?

**Comment by ZachJW34** [comment]
@Sayvai I'm hoping it is fixed now. Can you try one of the newer binaries (`https://cdn.cypress.io/beta/npm/10.3.1/darwin-x64/zachw/issue-22525-7a65ef80e1b2fd15f0abb743d53d80b61e9a1cdc/cypress.tgz`).

File: npm/webpack-dev-server/src/helpers/nextHandler.ts [hunk 2]
```diff
@@ -255,3 +247,52 @@ function sourceNextWebpack (devServerConfig: WebpackDevServerConfig, framework:
...
+        }
+      }
+    }
+  }
+}
+
+// Our modifications of the Next webpack config can corrupt the cache used for local development.
+// We separate the cache used for CT from the normal cache (".next/cache/webpack" -> ".next/cache/cypress-webpack") so they don't interfere with each other
+function changeNextCachePath (webpackConfig: Configuration) {
+  if (webpackConfig.cache && typeof webpackConfig.cache === 'object' && ('cacheDirectory' in webpackConfig.cache) && webpackConfig.cache.cacheDirectory) {
```
Comment (Sayvai) [inline npm/webpack-dev-server/src/helpers/nextHandler.ts:291]: 💡  I think you can actually simplify this conditional check with [Optional Chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining), like so:

```suggestion
  if (typeof webpackConfig?.cache === 'object' && ('cacheDirectory' in webpackConfig.cache) && webpackConfig.cache.cacheDirectory) {
```

**Comment by ZachJW34** [comment]
@Sayvai thanks for finding the issue and the words of encouragement! It goes a long way ❤️ 

Like you said, this should be available in the next release of Cypress!

**Review by mike-plummer — APPROVED** [review]
LGTM, just one potential comment/guard worth adding

**Review by ryanthemanuel — APPROVED** [review]
Looks good nice work!

File: npm/webpack-dev-server/src/helpers/nextHandler.ts [hunk 3]
```diff
@@ -255,3 +247,52 @@ function sourceNextWebpack (devServerConfig: WebpackDevServerConfig, framework:
...
+      ...webpackConfig.watchOptions,
+      ignored: [...webpackConfig.watchOptions.ignored.filter((pattern: string) => !/node_modules/.test(pattern)), '**/node_modules/!(@cypress/webpack-dev-server/dist/browser.js)**'],
+    }
+
+    debug('found options next.js watchOptions.ignored %O', webpackConfig.watchOptions.ignored)
+  }
+}
+
+const globalCssRe = [/(?<!\.module)\.css$/, /(?<!\.module)\.(scss|sass)$/]
+const globalCssModulesRe = [/\.module\.css$/, /\.module\.(scss|sass)$/]
```
[the remark this answers is not in the document]
Comment (ZachJW34) [reply npm/webpack-dev-server/src/helpers/nextHandler.ts:265]: That's fair, I'll add a comment and get this merged tomorrow morning if everything is green.

[the remark this answers is not in the document]
Comment (ZachJW34) [reply npm/webpack-dev-server/src/helpers/nextHandler.ts:269]: Addressed in ed808036a3dc85f7dfb1aa188aa8f315669ca9b1

**Pull request merged**